Use deltas in GScan

kinow opened the pull request:
These changes close a lot of issues, listed below. Do not worry about assigning to me, setting milestone, etc. After the final review, and when this issue is closed, then I will go through each issue linked below, will assign and set the milestone :+1: 

- close #542 (main issue)
- close #234
  * Now there is a JS object with the variables in `SubscriptionQuery` class, that gets filled from the VueRouter + VuexStore info via the `graphql.js` mixin :+1: 
  * **code review**
- close #209 
  * Now we declare the query and callbacks. Previously, there was some code in WorkflowService & GQuery that was calling setWorkflows(data.workflow), but now the callbacks receive the data as is, and can do whatever they need with it, be it nodes, workflows, deltas, etc.
  * **code review**
- close #160
  * Achieved that by using Mixins. Should be easy to port to Vue 3 too (thought about that when trying a few different approaches).
  * **code review**
- close #356 
  * **code review**
- partially address #376 (@oliver-sanders I've got an idea to improve it)
  * We are now adding views directly to the widgets. So any view (with the right mixins) can be loaded in the Workflow view & component.
  * **code review**
- close #388 
  * By coordinating in mixins when a subscription is merged/created, and only starting it in the VueRouter navigation guard (`subscriptionView` mixin) we guarantee that we don't start subscriptions that are immediately going to be stopped.
  * **review with browser network & other developer tools**
- close #416
  * We are not mixing the GScan and Tree queries any longer, they are kept as separated GraphQL web sockets subscriptions.
  * **code review or use browser network & other developer tools**
- close #453 
  * e2e tests added, also simplified with the mixins.
  * **code review**
- close #661 
  * The workflows data (GScan, Dashboard, WorkflowsTable) is in Vuex store's `workflows.workflows`. The workflow data is stored in `workflows.lookup` (workflow subscription lookup with `id` as key, and GraphQL data as value), and Tree view data is stored in `workflows.workflow`. The Tree view is re-using the data in `workflows.lookup` when it creates `workflows.workflow`, avoiding data duplication.
  * **code review and/or vue-dev-utils browser extension, inspecting the Vuex store when browsing different views**
- close #684 
  * See above.
- close #625
  * Noticed that after dismissing the alert, I was not getting notified about the new alerts (that I knew were supposed to happen). Remembered about this issue, and added a call to `setAlert(null)` when the user dismisses the alert. Note: if the user does not dismiss the alert, once a new alert appears, it will replace the existing alert :+1: 
  * **code review**
- close #688 
  * Depends.. it should close what was originally reported, but due to recent discussion, we may either remove the issue from this list, or create a follow-up issue.
  * **code review**
- close #671 
  * One more fixed in this branch :tada: was triaging issues, then decided to test this branch. A **query** does not have an initial data burst, so the UI kept waiting forever for GScan and Dashboard queries, loading until it got some data. But now as everything in the UI is using deltas after this branch (cc @hjoliver just FWIW, you can confirm by looking at the `queries.js` file in this branch, at what's exported… there should be only subscription queries with deltas) the issue does not happen anymore.

**Requirements check-list**
- [x] I have read `CONTRIBUTING.md` and added my name as a Code Contributor.
- [x] Contains logically grouped changes (else tidy your branch by rebase).
- [x] Does not contain off-topic changes (use other PRs for other changes).
- [x] Already covered by existing tests.
- [x] Appropriate change log entry included.
- [x] No documentation update required.

Notes from discussions on Element and comments here:

- [x] we can use `stateTotals` to retrieve the states in the workflow, with their count too
- [x] when showing states, we need to filter the ones that are applicable to job (i.e. filter with the `JobStatus` enum!)
- [x] we can handle when workflows are moved/renamed

kinow commented:
Ok, rebased, and now using `workflow.states` for the GScan workflow summary. It's using the task states, instead of job states. This causes some job icons added with the state "waiting", which is not valid for a job. So for now GScan may display an empty space (as the job component doesn't know what color to use).

It's also not showing any tooltip. But we will fix it before the beta release. Still need:

- [ ] check if we need tests, and if the tests are passing
- [ ] a final round of testing before it's ready for review

Then once it's merged, adding hierarchy to GScan should be a lot easier :+1:

kinow commented:
Fixed tests. Will squash some commits, and finish testing it on Monday morning :sleepy: :coffee:

kinow commented:
Almost there. Manual testing found two bugs. One was when you closed all the widgets. Now the second ... when you visit a workflow tree view, everything works OK. When you reload it, the "Add View" link disappears.

Tested `master` and it doesn't happen there, only on this branch :sweat:

kinow commented:
Hmm, found the reason. I had forgotten GScan was responsible for filling the Vuex global `workflows.workflows`. Now the Vue Router route finds a workflow name (e.g. `five`), and sets it in Vuex.

Then the Toolbar reacts to changes to that value. When we navigate to `/#/workflows/five`, it will get the workflow name and search the array of workflows in Vuex.

This was set by GScan before. It doesn't sound like the best way of doing it, but will fix it for now and think about how to improve it later.

kinow commented:
Fixed. Playing with the Several offline workflows, I found another bug (creating a separate issue for that), probably related to #454 I think.

@oliver-sanders this one should be ready to review. One important thing is to review the deltas subscription.

I have added a new subscription to `GScan`, and finished all the work here, just to realize that it could have gone into `WorkflowService`. Should I change that, and move both the `Workflow` view subscription, and the `GScan` component subscription there?

One thing to note is that normally the application will have a subscription ID `1` created when it creates `GScan`, and that that subscription will be running forever, as the `GScan` component is never destroyed :+1: 

And from what I remember, we are not able to merge deltas subscriptions. I used query fragments, but haven't found a solution to keep the existing query-merge from `gquery.js` for deltas.

oliver-sanders commented:
> And from what I remember, we are not able to merge deltas subscriptions.

Yeah, it gets messy. Conceptually it's easy:

```
subs:
    1: {a, b, {c}}
    2: {a: {d}}

incoming delta:
   {A, {C}}

internal deltas:
    1: {A, {C}}
    2: {A}
```

But in practice GraphQL isn't that abstract, need to spend some time on this one soonish to see if it's achievable easily.

oliver-sanders commented on src/components/cylc/gscan/GScan.vue:
```diff
@@ -158,70 +125,47 @@ export default {
+        if (added && added.workflow && added.workflow.status) {
+          const workflow = Object.assign({}, added.workflow, vm.workflows.get(id))
+          vm.workflows.set(workflow.id, workflow)
+          vm.workflowsTracker++
```
What is the purpose of `workflowsTracker`, I can't quite work it out.

kinow replied:
Vue 2 does not track reactivity of Set, Map... so you can either use a separate variable to trigger Vue's reactivity, or workaround with Vue.set, Vue.delete, maybe a watcher... in this case I thought this was the simplest solution.

And Vue3 fixes it, as you can have reactive Maps.

hjoliver replied:
Can you add a comment to the code, so we know what's going on here in future?

kinow replied:
There's one already where the variable is declared: https://github.com/cylc/cylc-ui/pull/543/files#diff-fbd6564546be7943d0780e9d928d7a79600e72699daed0fe29bbec1efae43a5fR95-R97

But I can add one here too, or try to improve that comment if necessary.

hjoliver replied:
Ah, OK, no worries.

kinow commented:
Fixing conflicts and rebasing in a few minutes.

kinow commented:
Fixing conflicts. But will need a little more time for making sure that filter & search work with the changes for the deltas 👍

kinow commented:
Should finish fixing the tests either today or Monday. But will definitely take at least until Monday for proper testing it before it's ready for review again, sorry.

kinow commented:
Unit tests should pass in the next CI run :crossed_fingers: . Now let me fix the e2e tests...

kinow commented:
Hopefully last commit should get lint, unit, e2e, and build, all passing. In the meantime, doing the last checks manually.

kinow commented:
One more commit, to prevent accessing `.states` from offline workflows. If CI passes, it should be ready for review. Finally :tada: :slightly_smiling_face:

oliver-sanders commented:
GScan doesn't seem to respond to the addition or removal of stopped workflows. I think it might not be responding to the full range of [flow change events](https://github.com/cylc/cylc-uiserver/blob/338b973e80050e7c77158b1bdf87980f067a27b1/cylc/uiserver/workflows_mgr.py#L192-L225). Perhaps we need to add more info to the deltas?

Steps to reproduce:

1. Create a workflow called `five` (but don't run it).
2. Load the UI, you should see it in the sidebar.
3. Rename `five` -> `six`.

The `five` flow should disappear and `six` should be added.

kinow commented:
Oh, never thought about renaming workflows. The UIS probably doesn't send a message telling the UIS to remove that workflow, or if does, I think the UI doesn't handle that :disappointed_relieved: 

The old code, that is not deltas-based, would handle that as it always re-rendered the whole list, so I think we can't push that to a follow-up. Will have to think about this one and confirm whether there's a solution in the UI. Otherwise will create and link an issue in UIS.

Thanks for the testing @oliver-sanders !

kinow commented:
Probably the fourth time rebasing a change with `GScan.vue` in it :smile: hopefully I did not remove too much, nor left too much when fixing the conflicts.

Needs:

- [ ] see if CI will pass, otherwise fix lint/unit/e2e tests
- [ ] see if the deltas are working OK
- [ ] start working/thinking on @oliver-sanders last comment :point_up:

kinow commented:
While fixing the e2e test, which doesn't look too hard to fix, noticed this in the console log:

![Screenshot from 2020-12-15 15-04-31](https://user-images.githubusercontent.com/304786/102158704-dd3b8e00-3ee6-11eb-89cd-d83a6d4e1ddb.png)

Trying to understand where that's happening, but not clear. Somewhere in the reactivity of `GScan`, I might have made an infinite loop by chaining two watcher/computed I think.

kinow commented:
Still struggling to see where the cyclic dependency is happening. So far that's what I think we have in this version of `GScan`.

![20201216_132658](https://user-images.githubusercontent.com/304786/102289003-88f5e400-3fa2-11eb-9580-4c7906216d98.jpg)

So

- `workflowTracker` is updated when a delta is applied to the `workflow` `Map` (not reactive, so we trick Vue by adding +1 to this var)
- `searchWorkflows` is updated when the user types something in the input field
- `filters` is updated when the user select a new task or workflow state filters
- `filteredWorkflows` is updated if any of the three above change, creating a filtered version of the `workflow` `Map`
- `sortedWorkflows` is updated if `filteredWorkflows` change, simply sorting the `filteredWorkflows` to display in the components templates

hjoliver commented:
> You may have an infinite update loop in a component render function

Trying to understand this. Is it: the render function is triggered reactively, then it updates the state of some data that triggers rendering again?

kinow commented:
> > You may have an infinite update loop in a component render function
> 
> Trying to understand this. Is it: the render function is triggered reactively, then it updates the state of some data that triggers rendering again?

Yes, that's my understanding too. Normally it's not too hard to spot where's the mistake. But I'm already preparing the debugger because so far just reading the code wasn't sufficient for my brain to understand what's going on :laughing:

kinow commented:
Found it. It wasn't in the JS code. It was in the HTML template. I had a `.sort()` in the array of states of a node. :man_facepalming: 

Now just need to fix one e2e.

kinow commented:
>GScan doesn't seem to respond to the addition or removal of stopped workflows. I think it might not be responding to the full range of flow change events. Perhaps we need to add more info to the deltas?

@dwsutherland I think this is something that is not being discussed anywhere else, and won't be fixed with the event-driven PR you are working on.

I think it is not possible with the deltas that we have at the moment. Or is there any way to find out when a workflow is deleted/renamed?

oliver-sanders commented:
I think that the required data is present in the deltas in the `workflow { status }` field.

In this example I do the following:

* Create a workflow
* Start it
* Hold it
* Release it
* Stop it
* Remove it

![state-changes](https://user-images.githubusercontent.com/16705946/102335815-4d4e2080-3f88-11eb-8231-261f1937945b.gif)

But there is something very strange going on. The workflow keeps showing up as added, even when it was only updated and sometimes the status is in the added part and sometimes it is in the updated part.

oliver-sanders commented:
Here's an example using `stripNull`:

```graphql
subscription {
  deltas(initialBurst: true) {
    id
    added(stripNull: true) {
      workflow {
        id
        status
      }
    }
    updated(stripNull: true) {
      workflow {
        id
        status
      }
    }
  }
}
```

in GraphiQL the "status" field in the updated delta keeps flickering, the status disappears and reappears regularly ever 8-10 seconds.

The `cylc scan` update is scheduled every 5 seconds so it seems unlikely that it would be the cause, I tried with this diff:

```diff --git a/cylc/uiserver/workflows_mgr.py b/cylc/uiserver/workflows_mgr.py
index 0a26aab..9fc415c 100644
--- a/cylc/uiserver/workflows_mgr.py
+++ b/cylc/uiserver/workflows_mgr.py
@@ -243,34 +243,44 @@ class WorkflowsManager:
         async for wid, before, after, flow in self._workflow_state_changes():
             # handle state changes
             if before == 'active' and after == 'inactive':
+                logger.debug(f'active => inactive: {wid}')
                 await self._disconnect(wid)
                 await self._stop(wid)
 
             elif before == 'active' and after is None:
+                logger.debug(f'active => None: {wid}')
                 await self._disconnect(wid)
                 await self._unregister(wid)
 
             elif before is None and after == 'active':
+                logger.debug(f'None => active: {wid}')
                 await self._register(wid, flow)
                 await self._connect(wid, flow)
 
             elif before is None and after == 'inactive':
+                logger.debug(f'None => inactive: {wid}')
                 await self._register(wid, flow)
 
             elif before == 'inactive' and after == 'active':
+                logger.debug(f'inactive => active: {wid}')
                 await self._connect(wid, flow)
 
             elif before == 'inactive' and after is None:
+                logger.debug(f'inactive => None: {wid}')
                 await self._unregister(wid)
 
             # finally update the new states for internal purposes
             if before == 'active':
+                logger.debug('#1')
                 self.active.pop(wid)
             elif before == 'inactive':
+                logger.debug('#2')
                 self.inactive.remove(wid)
             if after == 'active':
+                logger.debug('#3')
                 self.active[wid] = flow
             elif after == 'inactive':
+                logger.debug('#4')
                 self.inactive.add(wid)
 
     async def multi_request(self, command, workflows, args=None,
```

And confirmed that no unexpected state changes are occurring.

hjoliver commented:
Using @oliver-sanders example, I also see the workflow ID stay in **added** forever, and in **updated** _"status: running"_ flickers off briefly at intervals (every 3-20 seconds).

kinow commented:
I think the empty status could be an unintentional behaviour? I don't remember @dwsutherland or someone else intentionally setting it to empty to mean that the workflow had been removed. And not sure if that would still happen after the event-driven PR is merged.

hjoliver commented:
More than that, as these are *deltas* shouldn't the flow only appear in "added" once, and then in "updated" whenever its status changes?

hjoliver commented:
Ah, is it because the top level flow ID appears in "added" whenever a lower-level item (e.g. a new task) is added...

hjoliver commented:
From @oliver-sanders comment above:

> The workflow keeps showing up as added, even when it was only updated 

That's what I'm getting at too - "added" is supposed to show only *new* data elements, not what was new in the past and is now merely changing its value.

dwsutherland commented:
> in GraphiQL the "status" field in the updated delta keeps flickering, the status disappears and reappears regularly ever 8-10 seconds.

Probably because when it disappears the status hasn't been "updated" (but some other field has).. Sometimes it will have been set as a delta (even though the status hasn't actually changed, this will need to be changed at the suite)..

> I also see the workflow ID stay in added forever,

IDs will be sent with every delta... It's hard to say what other fields had changed under `workflow`, but usually `workflow` is only added once and then updated thereafter (unless the default arguments aren't what I think they are)..

Will have to have a play

hjoliver commented:
That makes sense, I think.

kinow commented:
Current GScan deltas query:

```graphql
subscription OnGscanDeltasData {
  deltas(stripNull: true) {
    ...GScanTreeDeltas
    __typename
  }
}

fragment GScanTreeDeltas on Deltas {
  id
  shutdown
  added {
    ...GscanAddedData
    __typename
  }
  updated {
    ...GscanUpdatedData
    __typename
  }
  __typename
}

fragment GscanAddedData on Added {
  workflow {
    ...WorkflowData
    __typename
  }
  __typename
}

fragment GscanUpdatedData on Updated {
  workflow {
    ...WorkflowData
    __typename
  }
  __typename
}

fragment WorkflowData on Workflow {
  id
  name
  status
  owner
  host
  port
  states
  __typename
}
```

Started JupyterHub, waited for the 2 subscriptions (dashboard/Gscan) to initiate, and copied the first message with the GScan deltas query :+1:  no variables needed.

oliver-sanders commented:
> Probably because when it disappears the status hasn't been "updated" (but some other field has).. Sometimes it will have been set as a delta (even though the status hasn't actually changed, this will need to be changed at the suite)..

Happy with it disappearing, don't like it reappearing, which it shouldn't do unless the value has changed.

I shifted the `id` into the `delas` field and used `stipNone`, on this platform I get a flicker every 1-3s:

![flicker](https://user-images.githubusercontent.com/16705946/102467800-a2516b80-4048-11eb-98da-4898b0cd4102.gif)

kinow commented:
What the tooltip is looking like now.

![GIFrecord_2020-12-18_120427](https://user-images.githubusercontent.com/304786/102554301-6215ea00-4129-11eb-926c-f7d7e689c36c.gif)

hjoliver commented:
Looks good!

kinow commented:
> Looks good!

I tried to make it as similar as possible to Cylc 7's. Under the summary, where it shows only the task name, in the future we can edit it to say "Recent running tasks" and then include the name of the recent tasks, once we have figured out the best way to implement that in the backend :+1: 

e2e tests failing as I re-generated the mocked offline data.

kinow commented:
Whew, CI passed. #559 is a trouble every time one needs to re-generate the offline data (i.e. whenever the GraphQL queries change)

kinow commented:
Rebasing this one today :nerd_face:

kinow commented:
Done. Had to delete and start from scratch. Even though I had re-installed everything, I had re-used an `venv` for Cylc UI Server, and apparently it refused to pick up the latest changes to the schema (no more `batchSysName`, now `jobRunnerName` I think).

I think @hjoliver had a similar issue, so without thinking twice re-created all the venvs, and this time it worked. We still have the blockers here like

- What about when a workflow is removed?
- What about when a workflow is renamed?
- Can we ignore the IDs generating new deltas even when nothing else in your query changed?

We already have the tests and changelog. So if we have the questions for these answers, this should be ready for review/merge :+1:

oliver-sanders commented:
https://matrix.to/#/!SDKYnTgYyOXNBpJySp:matrix.org/$161070481195180qBQbX:matrix.org?via=matrix.org

> What about when a workflow is removed?

I think we need a "pruned" delta from the UIS for this to work properly.

> What about when a workflow is renamed?

The UIS tracks flow "registrations" not flow content so renaming is a delete event, i.e. the UIS should provide these deltas:

* pruned(flow1)
* added(flow2)

kinow commented:
Rebased and did some manual testing. Same pending questions, no changes/fixes other than fixing conflicts.

kinow commented:
Hmmm, tried to rebase it but getting lost with so many changes to GScan from other PR's :sweat_smile: this is pending some discussion, so will leave this one to be rebased after that discussion, since there are other PR's for GScan pending review, that are more likely to be merged first anyway :+1:

kinow commented:
Picked the change that fixes the annoying `cylc-object` warnings in the console, when you run the tests, in another pull request #603 .

kinow commented:
Note to self: continue rebase tomorrow, & check how to fix `generate` (using `cylc install` && `cylc run`, need to be updated) and then fix the conflicts at `checkpoint.js` :nerd_face:

kinow commented:
> Note to self: continue rebase tomorrow, & check how to fix `generate` (using `cylc install` && `cylc run`, need to be updated) and then fix the conflicts at `checkpoint.js` nerd_face

Fixed. Before the `cylc install` command created a symlink, allowing the workflow to find the `get_checkpoint.py`. Now with the command installing it and using run names, the structure changed. I've changed the script to install without run names, and to copy the `get_checkpoint.py` into `~/cylc-run/one/bin/get_checkpoint.py`. `checkpoint.js` updated successfully, fixed conflicts, continuing with other conflicts.

kinow commented:
Rebased. The `build` check may pass (unit tests, linter, build), but I believe e2e will fail due to the `checkpoint` data missing some tasks like the ones under `GOOD` family. Will continue investigating later today/tomorrow :+1: :coffee:

kinow commented:
Even though some e2e tests are not passing, the latest commit fixes the last issue after the rebase. Deltas updating GScan. Had to disable the tasks summary until we figure out how we gonna do it with state totals, or calculating iterating tasks as before.

![image](https://user-images.githubusercontent.com/304786/110378681-bc846e00-80ba-11eb-9a3c-1254ba27ee80.png)

And we also still need to decide what to do when a workflow is deleted, renamed, etc.

kinow commented:
Rebased due to conflicts. e2e tests should still be failing.

kinow commented:
Rebased, build should still fail. But it's ready to soon use the work @dwsutherland showed yesterday with stateTotals + deltas. After this, this PR will also fix #401 (note to self: remember to link it here with closes ... later)

kinow commented:
Note to self: #627 is changing GScan's query, have to confirm that rebasing doesn't mess with the GraphQL fragments, and that the fragments are updated according to the latest GScan query.

kinow commented:
Starting to fix conflicts here. @dwsutherland did you think about a solution for the missing transition states in the UIS? Maybe a `state: deleted`? :grimacing: I'm probably oversimplifying how it would be fixed in the backend I think.

kinow commented on src/graphql/queries.js:
```diff
@@ -190,25 +218,39 @@ subscription {
+  }
+  updated {
+    ...GscanUpdatedData
+  }
```
Note to self: check if there's a pruned for gscan too? Can't recall if there was one and I ignored, or it simply was never there.

kinow replied:
Done.

The reason why I didn't include the `pruned` was that it didn't have one for workflows. Only families, jobs, tasks, and edges.

![image](https://user-images.githubusercontent.com/304786/113039569-4b853180-91f4-11eb-9131-41fe80826f78.png)

kinow reviewed (commented):
Leaving some review notes to myself. I will need to re-read the code in `master` and the changes here to remember what I have to discard after the `latestTaskstates`, and after our fixes for layout (wrapping in GScan) pre 0.3 release. :factory_worker:

kinow commented:
PR is up to speed with the latest development now. GScan showing workflows like in `master`, with no errors :tada: 

![image](https://user-images.githubusercontent.com/304786/113039813-969f4480-91f4-11eb-8311-767b506cab7e.png)

Pending only handling of removed/moved workflows :+1:

kinow commented:
Had a bit of fun this morning trying to send a `workflows{ pruned: ["kinow|abc"] }` to the UI. Modified the protobuf schema in flow, and a part of the resolvers to allow `pruned` in the query schema.

Then started setting breakpoints in the UIS data_store_mgr, to the point that the `pruned` delta is accepted, but for some reason it never makes it all the way to the resolver :laughing: looks like there's some special handling for workflows and for pruned nodes that I didn't manage to update correctly.

But it was still a good exercise to learn more about Cylc UIS & Flow interactions with the data stores, schemas, GraphQL, etc.

kinow commented:
Rebased. Will use this PR to review https://github.com/cylc/cylc-uiserver/pull/212

kinow commented:
Rebased. Tomorrow will start testing the UIS branches with this code again :+1:  might have to change the code a bit more to support the new workflow states, and changes in the deltas.

kinow commented:
Synced branches. Hadn't seen UIS PR had already been merged :sweat_smile: 

My current GraphQL query for GScan (added the `pruned.workflow` delta bit)

```graphql
subscription OnGscanDeltasData {
  deltas(stripNull: true) {
    ...GScanTreeDeltas
    __typename
  }
}

fragment GScanTreeDeltas on Deltas {
  id
  shutdown
  added {
    ...GscanAddedData
    __typename
  }
  updated {
    ...GscanUpdatedData
    __typename
  }
  pruned {
    workflow
    __typename
  }
  __typename
}

fragment GscanAddedData on Added {
  workflow {
    ...WorkflowData
    __typename
  }
  __typename
}

fragment GscanUpdatedData on Updated {
  workflow {
    ...WorkflowData
    __typename
  }
  __typename
}

fragment WorkflowData on Workflow {
  id
  name
  status
  owner
  host
  port
  stateTotals
  latestStateTasks(
    states: ["failed", "preparing", "submit-failed", "submitted", "running"]
  )
  __typename
}
```

The workflow is being deleted now. However, interestingly, in GraphiQL I get a delta with the added workflow when I add it back, but in JavaScript, the delta is **empty** :open_mouth: 

```js
{
  "id": "kinow|five",
  "shutdown": false,
  "added": {
    "__typename": "Added"
  },
  "updated": {
    "__typename": "Updated"
  },
  "pruned": {
    "__typename": "Pruned"
  },
  "__typename": "Deltas"
}
```

kinow commented:
Rebased.

kinow commented:
@hjoliver @oliver-sanders @dwsutherland I've added some code to handle the `deltas.pruned.workflow`. After syncing Cylc Flow, UIS, and checking out this branch, you should be able to take a look and see if GScan is behaving OK.

It should pass, hopefully, @oliver-sanders ' tests that failed last time (like deleting or renaming a workflow). If there are no issues found, then I'll add tests, squash commits, and prepare the code for a final review :tada: 

I noticed empty deltas in the deltas.workflow.added, and added `deltaStore:true` and `initialBurst: true` (see last commit), thanks to @dwsutherland . It solved the issue, but then it later occurred again. When looking at GraphiQL, I got the correct deltas. Not sure what's the cause, but @dwsutherland is aware and can help further investigating what's going on.

In the client side, you can add a `console.log(response)` [here](https://github.com/cylc/cylc-ui/pull/543/files#diff-fbd6564546be7943d0780e9d928d7a79600e72699daed0fe29bbec1efae43a5fR370) to look at the deltas returned, then play with the workflows and see what we are getting.

kinow commented on src/model/WorkflowState.model.js:
```diff
@@ -35,6 +36,8 @@ class WorkflowState extends Enumify {
   static STOPPED = new WorkflowState('stopped', mdiStopCircle)
   static ERROR = new WorkflowState('error', mdiHelpCircle)
+  static INSTALLED = new WorkflowState('installed', mdiFolderPlus)
+  static UNINSTALLED = new WorkflowState('uninstalled', null)
```
Note to self: remove this state, not in https://github.com/cylc/cylc-flow/blob/5e6ebe3cf1280e29fdc6b744d49271bb8f1cfec6/cylc/flow/workflow_status.py

hjoliver commented on src/components/cylc/gscan/GScan.vue:
```diff
@@ -280,13 +269,15 @@ export default {
-            }
-          })
+          items: WorkflowState.enumValues
+            .filter(state => state !== WorkflowState.UNINSTALLED)
```
Remove line? (spurious state now removed!)
```suggestion
```

hjoliver commented on src/views/Dashboard.vue:
```diff
@@ -232,20 +232,21 @@ export default {
+          return acc
+        }, {})
+      return WorkflowState.enumValues
+        .filter(state => state !== WorkflowState.UNINSTALLED)
```
Remove line?
```suggestion
```

dwsutherland commented:
I've had a look, and I can't reproduce the issue of the UI subscription:
```
subscription GscanSubscriptionQuery {
  deltas (stripNull: true, initialBurst: true) {
    ...GScanTreeDeltas
  }
}

# GSCAN DELTAS BEGIN

fragment GScanTreeDeltas on Deltas {
  id
  shutdown
  added {
    ...GscanAddedData
  }
  updated {
    ...GscanUpdatedData
  }
  pruned {
    workflow
  }
}

fragment GscanAddedData on Added {
  workflow (deltaStore: true) {
    ...WorkflowData
  }
}

fragment GscanUpdatedData on Updated {
  workflow (deltaStore: true) {
    ...WorkflowData
  }
}

fragment WorkflowData on Workflow {
  id
  name
  status
  owner
  host
  port
  stateTotals
  latestStateTasks(states: [
    "failed",
    "preparing",
    "submit-failed",
    "submitted",
    "running"
  ])
}
```
not getting the `installed` delta...

I've installed and removed the workflow so many times.. and all subscriptions get the `pruned.workflow`/`installed`

kinow commented:
Posted on Element, but just FWIW, I disabled the Dashboard subscription, and then the issue went away. Looks like you need two subscriptions at least to reproduce it :+1:

dwsutherland commented:
As mentioned on element it's a more general problem of the UI subscriptions (or at least the first two) sharing the same request, so have opened a new issue:
https://github.com/cylc/cylc-ui/issues/685

kinow commented:
Rebased.

oliver-sanders commented:
Hmm, all my workflows are showing as stopped at the moment:

<img width="450" alt="Screenshot 2021-06-10 at 12 48 03" src="https://user-images.githubusercontent.com/16705946/121519859-1d894e00-c9ea-11eb-96e8-f22fe9602294.png">

(a=stopped, b=paused, c=running)

oliver-sanders commented:
I'm getting some stuck workflow states in GScan.

Here's an example of a workflow which is paused. The Tree View can tell it is paused (the play button is showing and live data is showing) but GScan thinks it is installed (the folder icon is showing):

<img width="432" alt="Screenshot 2021-06-10 at 13 02 13" src="https://user-images.githubusercontent.com/16705946/121521543-1c592080-c9ec-11eb-81cd-49e98c9ddf89.png">

From GraphiQL query:

```json
{
  "data": {
    "workflows": [
      {
        "id": "oliver|one",
        "status": "paused"
      }
    ]
  }
}
```

Steps to replicate.

1. `cylc install <flow>`
2. `cylc play --pause <flow>`

kinow commented:
Something strange with reactivity. Found at least one bug in the `if/else` logic (used `if/else`, had to be separated `if`'s). I'm seeing stale states too, although the data appears to be corrected merged now. Will look at this tomorrow or next week.

Thanks for testing @oliver-sanders ! Looks like it at least removed the workflow correctly. I will add tests, squash, and find what's wrong. Next review should be the final one.

kinow commented on src/components/cylc/tree/deltas.js:
```diff
@@ -15,96 +15,62 @@
-  })
-}
+} from '@/components/cylc/tree/nodes'
+import * as CylcTree from '@/components/cylc/tree/index'
```
We cannot keep the state of the `CylcTree` in an object along with functions/methods. That's because we are now storing it in Vuex store, which doesn't allow functions/methods in the stored data :+1: The way the code is imported here, is very similar to how you can specify functions in a module in Python and call them directly. It looks like an Object, but is not really an instance with a constructor and local state.

kinow commented on src/components/cylc/tree/deltas.js:
```diff
@@ -181,71 +176,76 @@ function handleInitialDataBurst (deltas, tree) {
   // if added, removed, or updated deltas, we want to re-calculate the cycle point states now
   if (deltas.pruned || deltas.added || deltas.updated) {
-    tree.tallyCyclePointStates()
+    // FIXME
```
Note to self: remove this comment. I had commented the `tallyCyclePointStates()` call, and marked with a NOTE (the IDE highlights it, so it's easier for me to see what's pending), but forgot to remove the mark.

kinow commented on src/components/cylc/tree/deltas.js:
```diff
@@ -181,71 +176,76 @@ function handleInitialDataBurst (deltas, tree) {
+  // 2. data updated in deltas.updated
+  // 3. data pruned in deltas.pruned
+  try {
+    return handleDeltas(deltas, workflow, lookup, options)
```
**Note**: We are handling only deltas, no more handling initial data burst. I have modified the queries to use a flat structure, and whether it's the initial burst or not, makes no difference to the UI. I simply add every data returned to a dictionary and store in Vuex. Other Vuex actions (e.g. `applyTreeDeltas`) can then access the dictionary and use it to retrieve created values from delta, avoiding this way creating new - and duplicated - data.

kinow commented on src/components/cylc/tree/index.js:
```diff
@@ -14,58 +14,542 @@
+      Vue.delete(workflow[each], key)
+    })
+  })
+}
```
Lots of changes in this file. But you can mostly ignore them. The summary of what was changed is the following:

- No more `CylcTree` object that hold a `tree` structure and a `lookup`
- The functions of the `CylcTree` object were re-wrote into functions that accept the same parameters as before, plus a `workflow` parameter
- The `workflow` parameter comes from Vuex store; it has the following structure: `{ tree: {}, lookup: {}  }`

The `workflow.lookup` contains all the **nodes** of the tree in a flat hashmap. The `workflow.tree` contains the same data, but following a hierarchy. The `.node` of each entry in both `.lookup` and `.tree` points to the `workflows/lookup` global state (there are two lookups, maybe will rename one later).

kinow commented on src/components/cylc/tree/nodes.js:
```diff
@@ -332,7 +332,8 @@ function createCustomOutputs (job) {
 function createJobNode (job) {
-  Vue.set(job, 'customOutputs', createCustomOutputs(job))
+  const customOutputs = createCustomOutputs(job)
+  Vue.set(job, 'customOutputs', customOutputs)
```
Note to self: I had this change in my IDE when debugging, to see the value of `customOutputs` and confirm it was valid. This can probably be undone now.

kinow commented on src/components/cylc/workflow/Toolbar.vue:
```diff
@@ -165,14 +167,12 @@ export default {
       {
-        name: 'mutations',
-        title: 'Mutations',
+        name: MutationsView.name,
```
We don't need the lower-case and upper-case variations. Easier to fire an event just with the name and also use it for displaying the value. Plus, this way we just need to update the `.name` in one place to get the views&components that use it to update too.

kinow commented on src/components/cylc/workflow/deltas.js:
```diff
@@ -0,0 +1,130 @@
+  return {
+    errors
+  }
+}
```
This file contains the deltas for `workflow`. `workflows/workflow` is stored in Vuex, and used by GScan, Dashboard, and WorkflowTable component/views (for now, since the data structure is similar for all views/components, but that can be easily changed now).

kinow reviewed (commented):
Partially done a review for myself. Will continue work on Monday. But at the moment, this PR should address all feedback from previous reviews.

Plus, this should also fix several other issues that I will link on Monday. For instance:

- The flat-ish data is stored now in Vuex
- The subscriptions are kept in the Workflow Service
- Several new mixins are now used to control the lifecycle of subscriptions
- Workflow view simply adds new views; it does not care about the subscriptions

The size of the change will look intimidating once it's done (will have to add and update several tests), but I'm feeling quite confident that this will fix the GScan deltas and also allow us to incorporate the Table view PR more easily.

The reason for this big change, is that I realized that I would need yet another `deltasObservable` in the service for GScan. And that wouldn't look really nice once complete, and would have to be removed/fixed later when working on the Table PR.

More to come on Monday. I've hidden several of my previous comments to simplify the final review (which I may send the request next week) :+1:

oliver-sanders commented:
Tested workflow state transitions, looks good, works as expected.

kinow commented on src/views/Workflow.vue:
```diff
@@ -52,37 +52,31 @@ along with this program.  If not, see <http://www.gnu.org/licenses/>.
 </template>
 
 <script>
```
This file will have mainly code being removed. This is because the behavior of views & components with subscriptions has been standardized, and is now in mixins and in the Workflow Service. I will write a couple paragraphs on how to add a view with a widget here, and use that later to update the Table view PR.

kinow commented on src/store/workflows.module.js:
```diff
@@ -25,24 +58,88 @@ const getters = {
+  },
+  clearTree ({ commit }) {
+    commit('CLEAR_WORKFLOW')
   }
```
The data used in views with subscriptions is all here now.

We have new actions & mutations to both populate the data, and to remove it. These actions are called by the mixins & Workflow Service. You only need to declare what you want in your view or component.

kinow commented on src/services/workflow.service.js:
```diff
@@ -34,29 +27,38 @@ import {
    */
   constructor (httpUrl, subscriptionClient) {
-    super()
-    this.query = null
```
Before the Workflow Service allowed a single query. Every other query would be merged back to this original query. Now it supports multiple queries. You can have 1 merged query (e.g. when you have Dashboard + GScan), or you can have 2 or more (e.g. when you have GScan, and also the other query for Tree/Table views, which uses variables like `{ workflowId : "kinow|five" }`. Note these variables cannot be applied to the GScan query as that would limit the data returned, hence the 2 subs/queries in this example).

kinow commented on src/services/workflow.service.js:
```diff
@@ -106,108 +245,131 @@ class WorkflowService extends GQuery {
-      ),
-      this.apolloClient
-    )
+  gClone (query) {
```
This function, and the next functions down are all from the old `gquery.js`. I just moved them here. We are not changing how we merge queries, I've reverted that change, but will create disabled unit tests for cases where the current approach could result in bugs, or where it's missing some feature :+1: For now it should work fine for what we need.

kinow commented on src/services/mock/json/OnWorkflowTreeDeltasData.json:
```diff
@@ -1,366 +1,354 @@
 {
-  "workflows": [
+  "workflow": {
```
Flat data, same data that you should see in deltas!

kinow commented on src/router/index.js:
```diff
@@ -69,25 +70,35 @@ Vue.use(Meta)
+      .catch(error => {
+        const alert = new Alert(error, null, 'error')
+        return store.dispatch('setAlert', alert)
+      })
```
We already had a dependency/relationship between the router here and the store. Returning the user  in the User Service, eliminates the dependency between User Service and the store, making it easier to test.

kinow commented on src/model/ViewState.model.js:
```diff
@@ -0,0 +1,28 @@
+
+import { Enumify } from 'enumify'
+
+class ViewState extends Enumify {
```
Note to self: document.

kinow commented on src/model/SubscriptionQuery.model.js:
```diff
@@ -0,0 +1,38 @@
+// eslint-disable-next-line no-unused-vars
+import { DocumentNode } from 'graphql'
+
+class SubscriptionQuery {
```
Note to self: document.

kinow commented on src/model/Subscription.model.js:
```diff
@@ -0,0 +1,71 @@
+import ViewState from '@/model/ViewState.model'
+import Alert from '@/model/Alert.model'
+
+class Subscription {
```
Note to self: document

kinow commented on src/mixins/subscriptionView.js:
```diff
@@ -0,0 +1,46 @@
+  //     this.$workflowService.unsubscribe(this)
+  //   }
+  //   next()
+  // }
```
Note to self: TODO

kinow replied:
Aaaand document.

kinow commented on src/mixins/subscriptionComponent.js:
```diff
@@ -0,0 +1,30 @@
+
+import subscriptionMixin from '@/mixins/subscription'
+
+export default {
```
Note to self: document

kinow commented on src/mixins/subscription.js:
```diff
@@ -0,0 +1,35 @@
+import ViewState from '@/model/ViewState.model'
+import { mapActions } from 'vuex'
+
+export default {
```
Note to self: document.

kinow commented on src/layouts/Default.vue:
```diff
@@ -74,7 +73,7 @@ export default {
   errorCaptured (error, vm, info) {
     if (process.env.NODE_ENV !== 'production') {
-      store.dispatch('setAlert', new AlertModel(error, null, 'error'))
+      vm.$store.dispatch('setAlert', new AlertModel(error, null, 'error'))
```
Note to self: undo this change.

kinow commented on src/graphql/queries.js:
```diff
@@ -169,68 +108,130 @@ fragment JobData on Job {
   WORKFLOW_TREE_DELTAS_SUBSCRIPTION,
-  DASHBOARD_QUERY,
-  GSCAN_QUERY,
-  WORKFLOWS_TABLE_QUERY
```
Note to self: add the queries even though they won't result in a different `root` query?

kinow replied:
Dashboard added back.

kinow commented on tests/unit/components/cylc/tree/index.spec.js:
```diff
@@ -62,11 +80,11 @@ describe('Tree component functions', () => {
+    // const workflow = {}
+    // expect(populateTreeFromGraphQLData, null, workflow).to.not.throw(Error)
+    // expect(populateTreeFromGraphQLData, tree, null).to.not.throw(Error)
+    // expect(populateTreeFromGraphQLData, tree, workflow).to.not.throw(Error)
```
Note to self: undo?

kinow commented on src/components/cylc/tree/nodes.js:
```diff
@@ -343,24 +343,11 @@ function createJobNode (job) {
   createFamilyProxyNode,
   createTaskProxyNode,
   createJobNode,
-  containsTreeData,
```
Not used. Or, we could — in theory — use it to filter out empty deltas, see #689 But then TUI would have to do something similar. I think it would be best if we didn't receive any deltas after `shutdown: true`, unless it's a new initial burst.

kinow commented on src/components/core/Alert.vue:
```diff
@@ -56,8 +55,12 @@ export default {
       return this.colors.get(type) || ''
+    },
+    closeAlert () {
+      this.setAlert(null)
```
This is fixing #625 . See the `@click` in the Vuetify Alert component, which is bound to this `closeAlert` function, effectively removing the alert from the store too :+1: e2e test added for this.

kinow commented on src/components/core/Alert.vue:
```diff
@@ -56,8 +61,19 @@ export default {
+     * @param {Function} toggleFunction - the original Vuetify toggle function
+     * @see https://vuetifyjs.com/en/api/v-alert/
+     */
+    closeAlert (toggleFunction) {
```
This is fixing #625 . See the `@click` in the Vuetify Alert component, which is bound to this `closeAlert` function, effectively removing the alert from the store too :+1: e2e test added for this.

(`toggleFunction` is the `toggle` function slot prop from the Vuetify docs, which was called before to remove the Alert component from the UI.... we still call that, but first we remove the alert from the Vuex store...)

kinow commented:
@hjoliver @oliver-sanders this is now ready for review :tada: 

I linked the issues fixed in this PR at the top-most comment, adding a comment to each list item about where the review would happen. For most, you will have to read the code. But some issues can be reviewed running this branch with a running workflow, and using the browser tools & extensions.

There's an error message added for every cycle point in my `five` workflow. That's because the job finishes too soon, and the delta includes it in added, updated, and in the pruned sections. See Protocols element room on-going chat with @dwsutherland about it. I can add extra code to hide that message, but TUI would have to do the same if it re-used a delta (as well as other users/clients of our GraphQL endpoint), or we could filter those entities from the deltas - **pending**

I tested the following:

- linter OK
- unit tests OK
- e2e tests OK (`mutation.js` was brittle, failing occasionally when it failed to locate a DOM element; fixed now :crossed_fingers: )
- code review (reviewed the code last week, then fixed tests this week, and yesterday did a final code review of my own changes & PR comments that I had added with "Note to self" :grimacing: )
- manual tests
  * `five` workflow installed, then started
  * used tree view for the `five` workflow
  * stopped it, GScan and tree appear to be OK
  * `cylc install` another workflow, GScan shows the new workflow as installed
  * `cylc play --pause $workflow` now GScan shows the workflow's new status correctly
  * `cylc stop $workflow`, also OK
  * `cylc clean $workflow` all good too
  * `cylc install` the same workflow again
  * rm -rf ~/cylc-run/$workflow, workflow disappears from GScan
  * Installed 4 workflows, having 2 playing (being paused too), GScan and the Tree components both showing the correct status as far as I could tell, and GraphiQL is correct too

I hid several old comments where I was saying things like "Rebased", or "Tests are fixed", etc. Just to make review simpler. But note that there are 2 comments of Codecov. To see the latest coverage, you need to click on the "Read more" link in GitHub to find Codecov's latest comment (see the edit history dropdown for the date, it should be a few hours/days old). Or access https://app.codecov.io/gh/cylc/cylc-ui/compare/543.

Thanks!

kinow commented:
Rebased after merging @oliver-sanders 's fix for the mutations directive. Also spotted a typo in `CHANGES.md` so included an extra small commit :+1:

oliver-sanders commented:
Did some heavy testing, no issues so far 🍾.

It is very satisfying watching the job states update in GScan:

![gscan](https://user-images.githubusercontent.com/16705946/123413144-501a7580-d5aa-11eb-8fba-21cc5ad1371c.gif)

oliver-sanders commented:
Code is looks really good, I like the mixins, made it most the way through today, will come back to it Monday.

kinow commented on CHANGES.md:
```diff
@@ -10,14 +10,16 @@ release.
 of cycle points.
 
+[#543](https://github.com/cylc/cylc-ui/pull/543) - Use deltas in GScan.
+
```
This pull request solves several issues. After this gets merged, I will go through each issue updating it, and then create another PR with the changelog. I'm not sure if we won't have to drop some changes, so I didn't want to include the text here now and then maybe have to remove it later.

oliver-sanders reviewed (commented):
Made it to the end 😓.

Looks great, the queries and subs are nicely centralised and the mixins should make view development easier :+1:.

Still need to dive into the subscription merging (unexpected bonus 🎉). To help me review this could you comment about what the subscription merging is supposed to merge and what it isn't (i.e. what subscriptions we should expect to have for different views).

Still need to do some hands on testing with:

* Lookup / tree structures.
* Subscription merging.
* Delta processing.

I have only found a couple of small snags so far which are to do with the life cycle of subscriptions.

Here's the diff I used for testing (you may need to correct me if my method is wrong):

```diff
diff --git a/src/services/workflow.service.js b/src/services/workflow.service.js
index 455f60e8..ccdf521c 100644
--- a/src/services/workflow.service.js
+++ b/src/services/workflow.service.js
@@ -203,6 +203,8 @@ class WorkflowService {
     } catch (e) {
       subscription.handleViewState(ViewState.ERROR, e)
     }
+
+    console.warn('# START', this.subscriptions)
   }
 
   /**
@@ -271,6 +273,7 @@ class WorkflowService {
       store.dispatch(actionName).then(() => {})
     })
     delete this.subscriptions[subscription.query.name]
+    console.warn('# STOP', this.subscriptions)
   }
 
   // --- Merging GraphQL queries...
```

> **Note:** This logs the subscriptions AFTER a subscription has been started or stopped.

And here are my results:

> **Experiment 1:** Move from the Dashboard to workflow page.
>
> The "root" subscription is cancelled and re-started causing GScan to flicker.

<img width="404" alt="Screenshot 2021-06-28 at 15 52 38" src="https://user-images.githubusercontent.com/16705946/123661948-e3fb7400-d82c-11eb-9881-683ddb905bf0.png">

> **Experiment 2:** Open new tree views
>
> (Note: to replicate this I had to start on the Dashboard page)
>
> The "root" subscription is canceled and re-started.

<img width="429" alt="Screenshot 2021-06-28 at 15 57 59" src="https://user-images.githubusercontent.com/16705946/123661952-e52ca100-d82c-11eb-88ce-9b53ed9ebb20.png">

kinow commented:
Having some spare time this week after the work on this PR, and don't want to add more PR's that will cause this PR to have to be rebased or its merge delayed. So decided to cover a part of the new code that I thought wasn't too important to add tests initially (but doesn't hurt to have more tests right :) Last commit only includes a handful new unit tests, no changes to the code.

kinow commented on src/components/cylc/gscan/WorkflowIcon.vue:
```diff
@@ -42,9 +52,13 @@ export default {
-      return (
-        WorkflowState.enumValueOf(this.status.toUpperCase()) ||
-        WorkflowState.ERROR).icon
+      // TBD: enumValueOf returned the wrong value?
```
This is some crazy bug that I cannot explain at the moment. But using `enumValue for `WorkflowState` was returning the wrong value for me :confused: 

Will test again to confirm it was not some reactivity error (I hit this bug before I had the final version of this PR).

kinow replied:
Still happening, and no idea why that's happening. Could be an error when webpack or babel are processing the code, but will have to look at that later.

https://github.com/cylc/cylc-ui/issues/695

kinow commented on src/components/cylc/tree/deltas.js:
```diff
@@ -116,60 +82,96 @@ const UPDATED = {
- * @property {?DeltasPruned} pruned
+ * Helper object used to iterate pruned deltas data.
  */
+const PRUNED = {
+  jobs: 'removeJob',
+  taskProxies: 'removeTaskProxy',
+  familyProxies: 'removeFamilyProxy'
+}
```
I mentioned that to @dwsutherland yesterday on Element. But at the moment we are not handling when the workflow is removed in the tree. This code is being moved, but I agree we probably need to handle the removed. To be done in a follow-up, likely the #689 since I will be working on the deltas again.

oliver-sanders replied:
No probs, so long as it's on a ticket somewhere.

kinow commented on src/components/cylc/workflow/Toolbar.vue:
```diff
@@ -165,14 +167,12 @@ export default {
-        name: 'tree',
-        title: 'Tree',
+        name: TreeView.name,
         icon: mdiFileTree
```
Ah, this is related to the #376 issue. I left a comment at the top for you about it. I started to fix that here, but the change was getting a bit bigger. So I left only the initial work in this PR to simplify it, and moved the diff of what I had to this comment: https://github.com/cylc/cylc-ui/issues/376#issuecomment-868001793

Is that related to your suggestion about? WDYT?

oliver-sanders replied:
> Is that related to your suggestion about?

Yep

kinow replied:
To be fixed in #376

kinow commented on src/model/Subscription.model.js:
```diff
@@ -0,0 +1,85 @@
+    } else {
+      Object.values(this.subscribers).forEach(function (subscriber) {
+        subscriber.viewState = viewState
+        subscriber.setAlert(new Alert(context.message, null, 'error'))
```
The issue where we were showing only 1 error has been fixed here. Now it shows every alert added to the store.

A new alert replaces the last one. At the moment, you probably won't see all the alerts as Vue & the browser's DOM both batch changes like that. The alert component should display the last error in the subscribed view, but every error should be logged to the console (`setAlert` calls the Vuex store that calls `console`)

kinow replied:
For a follow-up, I think it would be nice if the user could see all the recent errors in some place in the UI, without having to open the browser console. Or maybe have a function that exports all the alerts (or with a limit of last-10, in case the user has like thousands of errors logged) so that that can be reported back to us.

kinow commented on src/services/workflow.service.js:
```diff
@@ -106,108 +245,131 @@ class WorkflowService extends GQuery {
+
+    subscribers.slice(1)
+      .forEach(subscriber => {
+        if (!isEqual(subscriber.query.variables, baseSubscriber.query.variables)) {
+          throw new Error('Error recomputing subscription: Query variables do not match.')
+        }
```
Ah, gotcha. I have a few post-its with questions about the merging of graphql queries, including the variables. There are other cases like directives and arguments that we need to decide what's the strategy for merging in a query, or if we start filtering... it'll be fun to work on this :nerd_face: 

Will keep that in mind :+1:

kinow commented on src/views/Workflow.vue:
```diff
@@ -130,70 +111,48 @@ export default {
-      } else if (view === 'mutations') {
-        Vue.set(this.widgets, (new Date()).getTime(), MutationsView.name)
-      } else {
+      switch (view) {
+      case TreeView.name:
+        Vue.set(this.widgets, createWidgetId(), TreeView.name)
+        break
+      case MutationsView.name:
+        Vue.set(this.widgets, createWidgetId(), MutationsView.name)
+        break
+      default:
         throw Error(`Unknown view "${view}"`)
       }
```
@oliver-sanders I think the `name` would end up always being the `MutationsView.name` name? Planning to fix this as a follow-up here: https://github.com/cylc/cylc-ui/issues/376#issuecomment-868001793 WDYT?

oliver-sanders replied:
Ah, whoops, copy and paste error, I meant to use the string argument provided to the function:

```diff
-Vue.set(this.widgets, createWidgetId(), MutationsView.name)
+Vue.set(this.widgets, createWidgetId(), view)
```

Fine to do later.

kinow replied:
Ah, perfect, that's exact what I did in the diff :+1:

kinow commented on src/views/Workflow.vue:
```diff
@@ -35,8 +35,8 @@ along with this program.  If not, see <http://www.gnu.org/licenses/>.
         >
-          <tree-component
-            :workflows="tree.root.children"
+          <tree-view
+            :workflow-name="workflowName"
           />
         </v-skeleton-loader>
```
Yup! Done here: https://github.com/cylc/cylc-ui/issues/376#issuecomment-868001793

And in the diff I've used `<component :is ... >` too! So looks like we are thinking in a similar way :) I started doing that here, but the change was getting a bit bigger, and would require more unit/e2e tests, so I postponed it... but I will be honest that probably I should have done it earlier, as this PR is already a little bloated (sorry!).

What do you think if once it's merged I started working on these improvements and we review it again?

oliver-sanders replied:
Ah yes, I had seen that diff but forgotten :facepalm:, great!

kinow commented on src/views/Guide.vue:
```diff
@@ -163,8 +163,11 @@ along with this program.  If not, see <http://www.gnu.org/licenses/>.
 
 export default {
+  name: 'Guide',
+  mixins: [subscriptionViewMixin],
```
Ah, that's the not-so-nice part of this design. We need `subscriptionViewMixin` for the view if it contains any GraphQL subscription query.

For the Guide page, it shows GScan, which requires us to include this mixin. I couldn't find a better way… 

With these mixins, we should be able to update to Vue 3 with none or very minimal changes. But with Vue3 we should be able to replace it by the composition API. Maybe that will give us another way to define this that will be more elegant. :disappointed_relieved:

oliver-sanders replied:
> For the Guide page, it shows GScan, which requires us to include this mixin. I couldn't find a better way…

Yikes, I hadn't realised that!

I'm guessing we can't just use the mixin on the top-level App.

kinow replied:
Correct, we have at least 1 page that doesn't use any subscription, the 404 page. Not sure if we will have more pages like that.

kinow commented on src/views/Dashboard.vue:
```diff
@@ -232,68 +240,20 @@ export default {
+          return acc
+        }, {})
+      return WorkflowState.enumValues
+        .sort((left, right) => left.name.localeCompare(right.name))
```
We can do that way too.

When you define enums, they get a `enumOrdinal` number value automatically assigned. So we would just need to define the enums, in the order we want to see in the UI I think. Should I do it here or as follow-up?

oliver-sanders replied:
Can be follow up if it detracts from this PR, not critical but good for usability.

kinow replied:
Created follow-up issue https://github.com/cylc/cylc-ui/issues/693

kinow commented on src/store/workflows.module.js:
```diff
@@ -25,24 +54,88 @@ const getters = {
+  applyWorkflowsDeltas ({ commit, state }, data) {
+    // modifying state directly in an action results in warnings...
+    const workflows = Object.assign({}, state.workflows)
+    applyDeltasGScan(data, workflows)
```
For now yes. But maybe it should be named `applyDeltasWorkflows`? WDYT?

kinow replied:
Will update it tomorrow (also need to look at a TODO for an enum that I couldn't finish today)

kinow replied:
Renamed!

kinow commented on src/store/workflows.module.js:
```diff
@@ -14,9 +14,38 @@
+import { clear } from '@/components/cylc/tree/index'
 
 const state = {
+  /**
+   * @type {Object.<String, Object>}
+   */
+  lookup: {},
```
Oh, my bad. I had a checkbox in one of the hidden comments to add comments, and then I only added a comment with the type :man_facepalming: . Let me fix this.

kinow commented on src/store/workflows.module.js:
```diff
@@ -14,9 +14,38 @@
+   */
+  workflow: {
+    tree: {},
+    lookup: {}
```
Ah, mea culpa too, I'm horrible naming things.

The `workflows/lookup` is the workflow view lookup used to look up GraphQL raw data. The idea is for components like Tree & Table to re-use the data in `workflows/lookup` as pointers, instead of duplicating it.

The `workflows/workflow` holds the data for a Tree view. Its`workflows/workflow.tree` contains just a single root node (workflow) that contains children data (hierarchical data).

And `workflows/workflow.lookup` is used as a lookup of tree nodes. When you add a new node to the tree, instead of using the `node.children`, we just grab the parent node from `workflows/workflow.lookup` and add the new node as child.

The new node will have `{ id: bla, node: $POINTER_TO_WORKFLOWS_LOOKUP_DATA, children: [] }`.

Hope this makes sense, and happy to update these names (just don't trust me to choose better names :laughing: )

oliver-sanders replied:
That makes sense, thanks for explaining. I think those names are great, just getting my head around the distinction between the `workflows` and `workflow` levels.

I guess with the hierarchical GScan we would add another `tree` structure at the `workflows` level.

kinow replied:
>I guess with the hierarchical GScan we would add another tree structure at the workflows level.

Yup, probably a very similar structure (since GScan re-uses the Tree component) :+1:

kinow commented on src/components/cylc/tree/index.js:
```diff
@@ -14,58 +14,548 @@
+ * @param {*} options
+ */
+function addCyclePoint (cyclePoint, workflow, options) {
+  const cyclePointsOrderDesc = options.cyclePointsOrderDesc !== undefined
+    ? options.cyclePointsOrderDesc
+    : DEFAULT_CYCLE_POINTS_ORDER_DESC
```
>Can move inside the if block below.

Good catch! done!

>Should this be applied at render rather than at tree build as I'm not sure how stable this will be if the user changes the preference?

Good question. At the moment it's implemented in a way where users have to visit the settings page, and reload the page with a tree view (if already open). Created an issue to track that: https://github.com/cylc/cylc-ui/issues/694

kinow commented on src/components/cylc/gscan/GScan.vue:
```diff
@@ -470,15 +416,48 @@ export default {
+        })
+      const summary = workflowStates
+        .map(stateTotal => `${stateTotal[0]}, ${stateTotal[1]}`)
+        .join(',')
```
Ops, fixed! Thanks!

kinow commented on src/services/workflow.service.js:
```diff
@@ -67,17 +69,154 @@ class WorkflowService extends GQuery {
+        {
+          next: function next (response) {
+            subscription.actionNames.forEach(actionName => {
+              store.dispatch(actionName, response.data).then(() => {})
```
Hmm, I wonder if I didn't speak too fast about fixing the issue of loading forever when there are no workflows.

Re-reading the code, now I can see that I'm changing the state immediately after I create the subscription, before it receives the first message where you pointed.

@oliver-sanders I can move the `subscription.handleViewState(ViewState.COMPLETE, null)` call here. The only downside I can see is that it would call the Vuex code for every subscription, which would be a bit expensive if you have a lot of GraphQL messages... :thinking:

oliver-sanders replied:
I'm not worried about moving the state change, was just wondering if something had been missed.

> Hmm, I wonder if I didn't speak too fast about fixing the issue of loading forever when there are no workflows.

Have re-tested, no probs.

kinow commented:
>Here's the diff I used for testing (you may need to correct me if my method is wrong):

@oliver-sanders I think your diff is correct. And the results of your test look correct to me. Initially, I re-used the GScan query for Dashboard & WorkflowTables. Which resulted in less subscriptions being stopped and started again, but then I thought it would be best to keep things as they are for now, and focus on getting the GScan component and the subscriptions in place first, leaving a few teaks like this one for later.

Probably to be addressed when we work on the new issues for query merging after this PR.

kinow commented on src/components/cylc/gscan/GScan.vue:
```diff
@@ -296,6 +293,8 @@ export default {
               value: state,
               model: false
             }
+          }).sort((left, right) => {
+            return left.text.localeCompare(right.text)
```
Ah, good idea! Done!

kinow commented on src/services/mock/generate:
```diff
@@ -22,8 +22,10 @@ fi
-cylc reg "one" one
-if cylc run "one" --fcp=20000102T0000Z --no-detach; then
+cylc install --flow-name="one" --no-run-name --directory=one
+mkdir "${HOME}/cylc-run/one/bin"
+cp checkpoint/get_checkpoint.py "${HOME}/cylc-run/one/bin/"
```
Is it `one/bin/get_checkpoint.py`, or `one/get_checkpoint.py`?

oliver-sanders replied:
`one/bin/get_checkpoint.py` (Cylc automatically adds the `bin/` directory to `$PATH`).

oliver-sanders commented:
Looking back at the issues I mentioned above about subscriptions being cancelled and re-started.

Perhaps the solution to this is to make the "unregister subscription" logic schedule a callback rather than cancelling the sub straight away. The callback can then check whether any new view has subscribed to the sub in the mean time.

This way the GScan subscription wouldn't get cancelled. This leads onto the idea of keeping workflow subs active for a configured period so that users can switch back and forth between workflows with minimum waiting. Related to https://github.com/cylc/cylc-ui/issues/392

kinow commented:
>looks like a bug on this branch causing the root sub to be cancelled when trying to add a workflow sub?

I thought the root subscription had been cancelled because the Dashboard query had been removed, resulting in a call to `recompute` and then triggering the stop/start in the subscription.

I will take a look tomorrow, but you are correct, it's supposed to stop `root` only iff the `root`'s query changed, not because of any changes to the `workflow` query.

kinow commented:
@oliver-sanders found the bug that was causing the GScan `root` subscription to be stopped and started again! The subscription was marked as `reload = true` because of the merge of GScan + Dashboard queries.

The issue was that I was never setting `reload = false` :sweat_smile: 

Fixed, and found another bug where I was not saving the result of union's of collections. Fix is really small, but I will probably need a bit more of time to write tests for these 2 issues I just found, to prevent them from happening again :+1: 

Thanks!!!

kinow commented:
Done! If tests pass in CI, it should not stop/start the root subscription when going from Dashboard to the Tree view or Workflow with Tree view :+1:

oliver-sanders commented:
Brilliant that fixes both of my scenarios.

oliver-sanders reviewed (approved):
Great :champagne:, really nice code, should give us a pattern to grow into.

kinow commented:
Thanks @oliver-sanders !

hjoliver commented:
I'm having a good play with this, probably can't improve on @oliver-sanders ' thorough review. Looks great so far, massive effort to close so many issues, thanks @kinow :bouquet:

hjoliver reviewed (approved):
Very nice :+1:
